### ISSUE TITLE
Bass Mix is missing

### ISSUE THREAD

**Issue by FadyQJ** [KernelErr0r/bass-sys#10]
The library is missing BASS_Mix bindings
Is it just a matter of adding 
```
binding BASS_MIXER_STREAMCREATE fn BASS_Mixer_StreamCreate(
        frequency: DWORD,
        channels: DWORD,
        flags: DWORD,
        proc: *mut STREAMPROC,
        user: *mut c_void
    ) -> HSTREAM;
```
? if so I can try to help

**Comment by FadyQJ** [comment]
It worked I added:
Functions.rs (added a temporary BASS_MIX_LIBRARY variable)
```
binding BASS_MIXER_GET_VERSION fn BASS_Mixer_GetVersion() -> DWORD;
    binding BASS_MIXER_STREAM_CREATE fn BASS_Mixer_StreamCreate(
        frequency: DWORD,
        channels: DWORD,
        flags: DWORD
    ) -> HSTREAM;
    binding BASS_MIXER_STREAM_ADD_CHANNEL fn BASS_Mixer_StreamAddChannel(
        handle: HSTREAM,
        channel: DWORD,
        flags: DWORD
    ) -> BOOL;
    binding BASS_MIXER_STREAM_ADD_CHANNEL_EX fn BASS_Mixer_StreamAddChannelEx(
        handle: HSTREAM,
        channel: DWORD,
        flags: DWORD,
        start: QWORD,
        length: QWORD
    ) -> BOOL;
    binding BASS_MIXER_STREAM_GET_CHANNELS fn BASS_Mixer_StreamGetChannels(
        handle: HSTREAM,
        channels: *mut DWORD,
        count: DWORD
    ) -> DWORD;
    binding BASS_MIXER_CHANNEL_GET_MIXER fn BASS_Mixer_ChannelGetMixer(
        handle: DWORD
    ) -> HSTREAM;
    binding BASS_MIXER_CHANNEL_IS_ACTIVE fn BASS_Mixer_ChannelIsActive(
        handle: DWORD
    ) -> DWORD;
    binding BASS_MIXER_CHANNEL_FLAGS fn BASS_Mixer_ChannelFlags(
        handle: DWORD,
        flags: DWORD,
        mask: DWORD
    ) -> DWORD;
    binding BASS_MIXER_CHANNEL_REMOVE fn BASS_Mixer_ChannelRemove(
        handle: DWORD
    ) -> BOOL;
    binding BASS_MIXER_CHANNEL_SET_POSITION fn BASS_Mixer_ChannelSetPosition(
        handle: DWORD,
        pos: QWORD,
        mode: DWORD
    ) -> BOOL;
    binding BASS_MIXER_CHANNEL_GET_POSITION fn BASS_Mixer_ChannelGetPosition(
        handle: DWORD,
        mode: DWORD
    ) -> QWORD;
    binding BASS_MIXER_CHANNEL_GET_POSITION_EX fn BASS_Mixer_ChannelGetPositionEx(
        channel: DWORD,
        mode: DWORD,
        delay: DWORD
    ) -> QWORD;
    binding BASS_MIXER_CHANNEL_GET_LEVEL fn BASS_Mixer_ChannelGetLevel(
        handle: DWORD
    ) -> DWORD;
    binding BASS_MIXER_CHANNEL_GET_LEVEL_EX fn BASS_Mixer_ChannelGetLevelEx(
        handle: DWORD,
        levels: *mut f32,
        length: f32,
        flags: DWORD
    ) -> BOOL;
    binding BASS_MIXER_CHANNEL_GET_DATA fn BASS_Mixer_ChannelGetData(
        handle: DWORD,
        buffer: *mut c_void,
        length: DWORD
    ) -> DWORD;
    binding BASS_MIXER_CHANNEL_SET_SYNC fn BASS_Mixer_ChannelSetSync(
        handle: DWORD,
        r#type: DWORD,
        param: QWORD,
        proc: *mut SYNCPROC,
        user: *mut c_void
    ) -> HSYNC;
    binding BASS_MIXER_CHANNEL_REMOVE_SYNC fn BASS_Mixer_ChannelRemoveSync(
        channel: DWORD,
        sync: HSYNC
    ) -> BOOL;
    binding BASS_MIXER_CHANNEL_SET_MATRIX fn BASS_Mixer_ChannelSetMatrix(
        handle: DWORD,
        matrix: *mut c_void
    ) -> BOOL;
    binding BASS_MIXER_CHANNEL_SET_MATRIX_EX fn BASS_Mixer_ChannelSetMatrixEx(
        handle: DWORD,
        matrix: *const c_void,
        time: f32
    ) -> BOOL;
    binding BASS_MIXER_CHANNEL_GET_MATRIX fn BASS_Mixer_ChannelGetMatrix(
        handle: DWORD,
        matrix: *mut c_void
    ) -> BOOL;
    binding BASS_MIXER_CHANNEL_SET_ENVELOPE fn BASS_Mixer_ChannelSetEnvelope(
        handle: DWORD,
        r#type: DWORD,
        nodes: *const BassMixerNode,
        count: DWORD
    ) -> BOOL;
    binding BASS_MIXER_CHANNEL_SET_ENVELOPE_POS fn BASS_Mixer_ChannelSetEnvelopePos(
        handle: DWORD,
        r#type: DWORD,
        pos: QWORD
    ) -> BOOL;
    binding BASS_MIXER_CHANNEL_GET_ENVELOPE_POS fn BASS_Mixer_ChannelGetEnvelopePos(
        handle: DWORD,
        r#type: DWORD,
        value: *mut f32
    ) -> QWORD;
    binding BASS_SPLIT_STREAM_CREATE fn BASS_Split_StreamCreate(
        channel: DWORD,
        flags: DWORD,
        chanmap: *const i32
    ) -> HSTREAM;
    binding BASS_SPLIT_STREAM_GET_SOURCE fn BASS_Split_StreamGetSource(
        handle: HSTREAM
    ) -> DWORD;
    binding BASS_SPLIT_STREAM_GET_SPLITS fn BASS_Split_StreamGetSplits(
        handle: DWORD,
        splits: *mut HSTREAM,
        count: DWORD
    ) -> DWORD;
    binding BASS_SPLIT_STREAM_RESET fn BASS_Split_StreamReset(
        handle: DWORD
    ) -> BOOL;
    binding BASS_SPLIT_STREAM_RESET_EX fn BASS_Split_StreamResetEx(
        handle: DWORD,
        offset: DWORD
    ) -> BOOL;
    binding BASS_SPLIT_STREAM_GET_AVAILABLE fn BASS_Split_StreamGetAvailable(
        handle: DWORD
    ) -> DWORD;
```

in Consts.rs
```
// BASS_Mixer_StreamCreate flags
pub const BASS_MIXER_RESUME: DWORD = 0x1000;	// resume stalled immediately upon new/unpaused source
pub const BASS_MIXER_POSEX: DWORD = 0x2000;	// enable BASS_Mixer_ChannelGetPositionEx support
pub const BASS_MIXER_NOSPEAKER: DWORD = 0x4000;	// ignore speaker arrangement
pub const BASS_MIXER_QUEUE: DWORD = 0x8000;	// queue sources
pub const BASS_MIXER_END: DWORD = 0x10000;	// end the stream when there are no sources
pub const BASS_MIXER_NONSTOP: DWORD = 0x20000;	// don't stall when there are no sources

// BASS_Mixer_StreamAddChannel/Ex flags
pub const BASS_MIXER_CHAN_ABSOLUTE: DWORD = 0x1000;	// start is an absolute position
pub const BASS_MIXER_CHAN_BUFFER: DWORD = 0x2000;	// buffer data for BASS_Mixer_ChannelGetData/Level
pub const BASS_MIXER_CHAN_LIMIT: DWORD = 0x4000;	// limit mixer processing to the amount available from this source
pub const BASS_MIXER_CHAN_MATRIX: DWORD = 0x10000;	// matrix mixing
pub const BASS_MIXER_CHAN_PAUSE: DWORD = 0x20000;	// don't process the source
pub const BASS_MIXER_CHAN_DOWNMIX: DWORD = 0x400000; // downmix to stereo/mono
pub const BASS_MIXER_CHAN_NORAMPIN: DWORD = 0x800000; // don't ramp-in the start
pub const BASS_MIXER_BUFFER: DWORD = BASS_MIXER_CHAN_BUFFER;
pub const BASS_MIXER_LIMIT: DWORD = BASS_MIXER_CHAN_LIMIT;
pub const BASS_MIXER_MATRIX: DWORD = BASS_MIXER_CHAN_MATRIX;
pub const BASS_MIXER_PAUSE: DWORD = BASS_MIXER_CHAN_PAUSE;
pub const BASS_MIXER_DOWNMIX: DWORD = BASS_MIXER_CHAN_DOWNMIX;
pub const BASS_MIXER_NORAMPIN: DWORD = BASS_MIXER_CHAN_NORAMPIN;

// Mixer attributes
pub const BASS_ATTRIB_MIXER_LATENCY: DWORD = 0x15000;
pub const BASS_ATTRIB_MIXER_THREADS: DWORD = 0x15001;
pub const BASS_ATTRIB_MIXER_VOL: DWORD = 0x15002;

// Additional BASS_Mixer_ChannelIsActive return values
pub const BASS_ACTIVE_WAITING: DWORD = 5;
pub const BASS_ACTIVE_QUEUED: DWORD = 6;

// BASS_Split_StreamCreate flags
pub const BASS_SPLIT_SLAVE: DWORD = 0x1000;	// only read buffered data
pub const BASS_SPLIT_POS: DWORD = 0x2000;

// Splitter attributes
pub const BASS_ATTRIB_SPLIT_ASYNCBUFFER: DWORD = 0x15010;
pub const BASS_ATTRIB_SPLIT_ASYNCPERIOD: DWORD = 0x15011;

// Envelope types
pub const BASS_MIXER_ENV_FREQ: DWORD = 1;
pub const BASS_MIXER_ENV_VOL: DWORD = 2;
pub const BASS_MIXER_ENV_PAN: DWORD = 3;
pub const BASS_MIXER_ENV_LOOP: DWORD = 0x10000; // flag: loop
pub const BASS_MIXER_ENV_REMOVE: DWORD = 0x20000; // flag: remove at end

// Additional sync types
pub const BASS_SYNC_MIXER_ENVELOPE: DWORD = 0x10200;
pub const BASS_SYNC_MIXER_ENVELOPE_NODE: DWORD = 0x10201;
pub const BASS_SYNC_MIXER_QUEUE: DWORD = 0x10202;

// Additional BASS_Mixer_ChannelSetPosition flag
pub const BASS_POS_MIXER_RESET: DWORD = 0x10000; // flag: clear mixer's playback buffer

// Additional BASS_Mixer_ChannelGetPosition mode
pub const BASS_POS_MIXER_DELAY: DWORD = 5;

// BASS_CHANNELINFO types
pub const BASS_CTYPE_STREAM_MIXER: DWORD = 0x10800;
pub const BASS_CTYPE_STREAM_SPLIT: DWORD = 0x10801;
```

In Types.rs
```
// Envelope node
pub struct BassMixerNode {
    post: QWORD,
    value: f32
}

impl BassMixerNode {
    pub fn new(post: QWORD, value: f32) -> Self {
        Self { post, value }
    }
}
```

**Comment by thatbakamono** [comment]
I'd want to support bass mix and other addons but to be honest, bass-sys in the current state is not a good place for that for two reasons:
1) I'm not sure whether providing bindings for all those libraries in a single place is a good idea, I think it'd be a better idea to have separate crates for everything. In that case I'd probably switch to a new repository.
2) Everything is written by hand and supporting bass and 19 addons manually... Is just too tedious and error prone. I think I'd be better to switch to [bindgen](https://github.com/rust-lang/rust-bindgen) but that requires some work (in form of procedural macros) as bindgen's support for libloading is **very** limited.

Tl;dr I can do that but it's gonna take some time and is going to be in a new repository. If you want to help, it'd be probably best to reach me on Discord.

**Comment by FadyQJ** [comment]
Well I added the parts I was missing, and it is working right now.
My project will not be using Bass on release but I will just be using it temporarily.

If I did anything helpful meanwhile I will get in touch with you, but I cannot support otherwise due to my very limited time schedule. However, I may delegate this project to someone whom I am teaching as practice for wrapping libraries.
